### PR TITLE
Add soft quadrupole

### PR DESCRIPTION
**Pull request by cemitch99**
Implementation of quadrupole w/ soft-edge fringe fields.

Math for map computation, default coefficients, and example parameters have been updated.

- [x] Modify composite maps used for numerical integration
- [x] Modify default on-axis coefficients
- [x] Modify example inputs & docs

### CODE REVIEW

File: examples/quadrupole_softedge/README.rst [hunk 1]
```diff
@@ -0,0 +1,54 @@
+.. _examples-solenoid-softedge:
+
+Soft-edge solenoid
+===================
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]:
```suggestion
Soft-Edge Solenoid
==================
```

File: examples/quadrupole_softedge/README.rst [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+We use a 250 MeV proton beam with initial unnormalized rms emittance of 1 micron
+in the horizontal plane, and 2 micron in the vertical plane.
+
+In this test, the initial and final values of :math:`\sigma_x`, :math:`\sigma_y`, :math:`\sigma_t`, :math:`\epsilon_x`, :math:`\epsilon_y`, and :math:`\epsilon_t` must agree with nominal values.
+
+
+Run
+---
+
+This example can be run as a Python script (``python3 run_solenoid_softedge.py``) or with an app with an input file (``impactx input_solenoid_softedge.in``).
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]:
```suggestion
This example can be run as a Python script (``python3 run_quadrupole_softedge.py``) or with an app with an input file (``impactx input_quadrupole_softedge.in``).
```

File: examples/quadrupole_softedge/README.rst [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+       .. literalinclude:: run_solenoid_softedge.py
+          :language: python3
+          :caption: You can copy this file from ``examples/solenoid_softedge/run_solenoid_softedge.py``.
+
+   .. tab-item:: App Input File
+
+       .. literalinclude:: input_solenoid_softedge.in
+          :language: ini
+          :caption: You can copy this file from ``examples/solenoid_softedge/input_solenoid_softedge.in``.
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]:
```suggestion
       .. literalinclude:: run_quadrupole_softedge.py
          :language: python3
          :caption: You can copy this file from ``examples/quadrupole_softedge/run_quadrupole_softedge.py``.

   .. tab-item:: App Input File

       .. literalinclude:: input_quadrupole_softedge.in
          :language: ini
          :caption: You can copy this file from ``examples/quadrupole_softedge/input_quadrupole_softedge.in``.
```

File: examples/quadrupole_softedge/README.rst [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+Analyze
+-------
+
+We run the following script to analyze correctness:
+
+.. dropdown:: Script ``analysis_solenoid_softedge.py``
+
+   .. literalinclude:: analysis_solenoid_softedge.py
+      :language: python3
+      :caption: You can copy this file from ``examples/solenoid_softedge/analysis_solenoid_softedge.py``.
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]:
```suggestion
.. dropdown:: Script ``analysis_quadrupole_softedge.py``

   .. literalinclude:: analysis_quadrupole_softedge.py
      :language: python3
      :caption: You can copy this file from ``examples/quadrupole_softedge/analysis_quadrupole_softedge.py``.
```

File: examples/quadrupole_softedge/input_quadrupole_softedge.in [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+# Beamline: lattice elements and segments
+###############################################################################
+lattice.elements = quad1
+lattice.nslice = 1
+
+quad1.type = quadrupole_softedge
+quad1.ds = 6.0
+quad1.gscale = 1.233482899483985
+quad1.mapsteps = 800
+
```
Comment (ax3l) [inline examples/quadrupole_softedge/input_quadrupole_softedge.in:63]:
```suggestion


```

File: src/python/elements.cpp [hunk 6]
```diff
@@ -245,4 +245,21 @@ void init_elements(py::module& m)
...
     ;
+
+    py::class_<SoftQuadrupole, elements::Thick>(me, "SoftQuadrupole")
+        .def(py::init<
+                 amrex::ParticleReal const,
+                 amrex::ParticleReal const,
+                 std::vector<amrex::ParticleReal>,
+                 std::vector<amrex::ParticleReal>,
+         int const,
+                 int const
```
Comment (ax3l) [inline src/python/elements.cpp]:
```suggestion
                 amrex::ParticleReal,
                 amrex::ParticleReal,
                 std::vector<amrex::ParticleReal>,
                 std::vector<amrex::ParticleReal>,
                 int,
                 int
```

**Comment by cemitch99** [comment]
All tests now pass for the soft quadrupole except CUDA / NVCC 11.0.2 SP.

File: docs/source/usage/parameters.rst [hunk 7]
```diff
@@ -225,6 +225,22 @@ Lattice Elements
...
             * ``<element_name>.nslice`` (``integer``) number of slices used for the application of space charge (default: ``1``)
 
+        * ``quadrupole_softedge`` for a soft-edge quadrupole. This requires these additional parameters:
+
+            * ``<element_name>.ds`` (``float``, in meters) the segment length
+
+            * ``<element_name>.gscale`` (``float``, in inverse meters) Scaling factor for on-axis magnetic field gradient
+
+            * ``<element_name>.cos_coefficients`` (array of ``float``) cos coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a thin-shell model from `DOI:10.1016/J.NIMA.2022.166706 <https://doi.org/10.1016/j.nima.2022.166706>`__
```
Comment (ax3l) [inline docs/source/usage/parameters.rst]: Discussed today: we are looking for a good reference we can use for a soft edge quad as data input.

If we don't find a good experimental model, we could use model coefficients after the hyberbolic tangent model used in MaryLie.

Comment (cemitch99) [reply docs/source/usage/parameters.rst]: Done.

File: docs/source/usage/parameters.rst [hunk 8]
```diff
@@ -225,6 +225,22 @@ Lattice Elements
...
+
+            * ``<element_name>.ds`` (``float``, in meters) the segment length
+
+            * ``<element_name>.gscale`` (``float``, in inverse meters) Scaling factor for on-axis magnetic field gradient
+
+            * ``<element_name>.cos_coefficients`` (array of ``float``) cos coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
+
+            * ``<element_name>.sin_coefficients`` (array of ``float``) sin coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```
Comment (ax3l) [inline docs/source/usage/parameters.rst]:
```suggestion
              (optional); default is a tanh fringe field model from `MaryLie 3.0 <http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```

File: docs/source/usage/parameters.rst [hunk 9]
```diff
@@ -225,6 +225,22 @@ Lattice Elements
...
             * ``<element_name>.nslice`` (``integer``) number of slices used for the application of space charge (default: ``1``)
 
+        * ``quadrupole_softedge`` for a soft-edge quadrupole. This requires these additional parameters:
+
+            * ``<element_name>.ds`` (``float``, in meters) the segment length
+
+            * ``<element_name>.gscale`` (``float``, in inverse meters) Scaling factor for on-axis magnetic field gradient
+
+            * ``<element_name>.cos_coefficients`` (array of ``float``) cos coefficients in Fourier expansion of the on-axis field gradient
+              (optional); default is a tanh fringe field model from `<http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```
Comment (ax3l) [inline docs/source/usage/parameters.rst]: Title of the link missing:
```suggestion
              (optional); default is a tanh fringe field model from `MaryLie 3.0 <http://www.physics.umd.edu/dsat/docs/MaryLieMan.pdf>`__
```

File: examples/quadrupole_softedge/README.rst [hunk 10]
```diff
@@ -0,0 +1,46 @@
+.. _examples-quadrupole-softedge:
+
+Soft-Edge Quadrupole
+===================
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]:
```suggestion
Soft-Edge Quadrupole
====================
```

File: examples/quadrupole_softedge/README.rst [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+This is a modification of the test for a matched electron beam propagating through a stable
+FODO cell, in which the quadrupoles have been replaced with soft-edge quadrupole
```
Comment (ax3l) [inline examples/quadrupole_softedge/README.rst]: Let's link the test we mean. You refer to `examples/fodo/`, right? :)
```suggestion
This is a modification of :ref:`the test for a matched electron beam propagating through a stable FODO cell <examples-fodo>`,
in which the quadrupoles have been replaced with soft-edge quadrupole
```

File: examples/quadrupole_softedge/input_quadrupole_softedge.in [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+quad1.cos_coefficients =
+          2.0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
+quad1.sin_coefficients =
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
```
Comment (ax3l) [inline examples/quadrupole_softedge/input_quadrupole_softedge.in:42]: Wouldn't this work, too?
```suggestion
quad1.cos_coefficients = 2.0
quad1.sin_coefficients = 0.0
```

File: examples/quadrupole_softedge/input_quadrupole_softedge.in [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+quad2.cos_coefficients =
+          2.0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
+quad2.sin_coefficients =
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0 0 0 0 0 0
+            0 0 0 0 0
```
Comment (ax3l) [inline examples/quadrupole_softedge/input_quadrupole_softedge.in:60]:
```suggestion
quad2.cos_coefficients = 2.0
quad2.sin_coefficients = 0.0
```

File: examples/quadrupole_softedge/run_quadrupole_softedge.py [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+    ],
```
Comment (ax3l) [inline examples/quadrupole_softedge/run_quadrupole_softedge.py:89]:
```suggestion
    cos_coefficients=[
        2,
    ],
```

File: examples/quadrupole_softedge/run_quadrupole_softedge.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
```
Comment (ax3l) [inline examples/quadrupole_softedge/run_quadrupole_softedge.py:125]:
```suggestion
```

File: examples/quadrupole_softedge/run_quadrupole_softedge.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
```
Comment (ax3l) [inline examples/quadrupole_softedge/run_quadrupole_softedge.py:169]:
```suggestion
        2,
```

File: examples/quadrupole_softedge/run_quadrupole_softedge.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
```
Comment (ax3l) [inline examples/quadrupole_softedge/run_quadrupole_softedge.py:206]:
```suggestion
```

File: src/particles/elements/SoftQuad.H [hunk 18]
```diff
@@ -0,0 +1,479 @@
...
+            amrex::ParticleReal pxout = px;
+            amrex::ParticleReal pyout = py;
+            amrex::ParticleReal ptout = pt;
+
+            // get the linear map
+            amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const R = refpart.map;
+
+            // symplectic linear map for a quadrupole is computed using the
+            // Hamiltonian formalism as described in:
+            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf.
```
Comment (ax3l) [inline src/particles/elements/SoftQuad.H]: Sometimes safer for links that people might copy:
```suggestion
            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf .
```

**Review by ax3l — COMMENTED** [review]
This is great, thank you!

I fixed the CUDA issue and added a few inline suggestions.

Feel free to adjust the PR description, it currently reads `Preliminary template only.`?

**Review by ax3l — APPROVED** [review]
Thank you, this is great!

I'll investigate separately why the different sized number of coefficients do not work for this and other elements.

**Comment by ax3l** [comment]
Ah, I think there is a bug in `ParmParse::queryAdd` and `ParmParse::queryarr` for `std::vector` arguments.
The passed array is only up-sized if needed, but not down-sized if the key exists.
This leads to bugs for long default values and shorter user-provided arrays.
https://github.com/AMReX-Codes/amrex/blob/23.03/Src/Base/AMReX_ParmParse.cpp#L877-L880

**Pull request merged**